docs/_posts/2015-02-24-streamlining-react-elements.md

```diff
@@ -0,0 +1,230 @@
+
+```js
+var element = this.props.child;
+element.props.count = this.state.count;
+return element;
+```
+
+You take a ReactElement through `props.child` and mutate its property before rendering it. If this component's state update, it won't actually get a new ReactElement in `props.child`. It will be the same one. You're mutating the same props.
+
+You could imagine that this would work. However, this disables the ability for any component to use `shouldComponentUpdate`. It looks like the component never changed because the previous value is always the same as the next one. Since the DOM layer does diffing, this pattern doesn't even work in this case. The change will never propagate down to the DOM except the first time.
+
+Additionally, if this element is reused in other places or used to switch back and forth between two modes, then you have all kinds of weird race conditions.
+
+It has always been broken to mutate the props of something passed into you. The problem is that we can’t warn you about this special case if you accidentally do this.
+
+### Problem: Too Late Validation
+
+In React 0.12, we do PropType validation very deep inside React during mounting. This means that by the time you get an error, the debugger stack is long gone. This makes it difficult to find complex issues during debugging. We have to do this since it is fairly common for extra props to be added between the call to React.createElement and the mount time. So the type is incomplete until then.
+
+The static analysis in Flow is also impaired by this. There is no convenient place in the code where Flow can determine that the props are finalized.
+
+### Solution: Immutable Props
+
+Therefore, we would like to be able to freeze the element.props object so that it is immediately immutable at the JSX callsite (or createElement). In React 0.13 we will start warning you if you mutate props after this point.
+
+You can generally refactor these pattern to simply use two different JSX calls:
+
+```js
+if (shouldUseFoo) {
+  return <Foo foo={10} bar={true} />;
+} else {
+  return <Foo bar={false} />;
+}
+```
+
+However, if you really need to dynamically build up your props you can just use a temporary object and spread it into JSX:
+
+```js
+var props = { bar: false };
+if (shouldUseFoo) {
+  props.foo = 10;
+  props.bar = true;
+}
+return <Foo {...props} />;
+```
+
+It is still OK to do deep mutations of objects. E.g:
+
+```js
+return <Foo nestedObject={this.state.myModel} />;
+```
+
+In this case it's still ok to mutate the myModel object in state. We recommend that you use fully immutable models. E.g. by using immutable-js. However, we realize that mutable models are still convenient in many cases. Therefore we will only shallow freeze the props object that belongs to the ReactElement itself. Not nested objects.
+
+### Solution: Early PropType Warnings
```

Comment:
* Problem 1
* Problem 2
* Solution 1
* Solution 2

Should we maybe order these differently? I this one is related to the immutability bits, but it still feels weird to present things in this order.